kpatch-build/create-diff-object.c

```diff
@@ -766,6 +766,12 @@ do {								\
 	e2->twin = e1;						\
 	/* set initial status, might change */			\
 	e1->status = e2->status = SAME;				\
+	if (strcmp(e1->name, e2->name)) {			\
+		/* Rename mangled element */			\
+		log_debug("renaming %s to %s\n",		\
+			  e2->name, e1->name);			\
+		e2->name = strdup(e1->name);			\
```

Comment:
The name of an element might point to a string in a elf file which might be closed when we still are using e2, so we need to dup (or not close the original elf nor the patched elf until we quit the program).

As for memory leaking, I must admit I followed the way the code Josh deleted used to handle things. The issue is that at this point a section or symbol name could point to memory mapped from an elf file or memory allocated through malloc.
Since kpatch is just a "one-off" program, having this memory allocated until the end of the program should be too much of a big deal. But if we want to avoid that we can strdup all names and free them in the kpatch_elf_teardown().

Let me know if we shoudl go with that.

Reply:
Well, I personally try to keep leaks to minimum even for "one-offs", but I did see a comment somewhere in the same vein of "we don't care since its a one-off", so I'll leave it for @jpoimboe to decide.

Reply:
I could only see two ways to solve this:
- Always use malloc for the name of symbols and sections, but that would make us always use more memory than we need to avoid a leak that the OS will cleanup once the program finishes.
- Have a way to distinguish whether the name was allocated or not and check at the end.

First solution doesn't seems a bit wasteful just to avoid a leak that isn't that much of an issue (in my opinion) and I would prefer to keep the leak.
Second solution can be a bit awkward, but after sleeping on it we could pretend to make the distinction a bit more "useful". We could have a field `orig_name` in section and symbol structures, which by default would stay `NULL`. And if an element gets renamed `orig_name` would point to the string in the ELF while `name` would point to freeable memory. So we can know there is a need to free `name` if `orig_name != NULL` or something like that. While there is no big need for keeping the original name, I can at least see some use to it for debugging (e.g. could access it through GDB) or logging purposes.

Otherwise we can also keep the leak.

Let me know which you prefer.

Reply:
Also, I just check, we already have the same leak when renaming static local variables.

Reply:
FWIW, some time in the past I wrapped kpatch-build with valgrind and tried to get the non-paths to be mostly silent.  For a userspace program, I wasn't going to pull my hair out getting all the various error paths to be perfect... but at the very least a clean successful build is helpful to find any new bad behavior.  /two-cents